db/migrate/20190801143210_convert_dns_conflict_timeout_setting.rb

```diff
@@ -0,0 +1,46 @@
+class FakeSetting < ApplicationRecord
+  self.table_name = 'settings'
+
+  def default=(v)
+    write_attribute :default, v.to_yaml
+  end
+
+  def value=(v)
+    v = v.to_yaml unless v.nil?
+    write_attribute :value, v
+  end
+
+  def value
+    v = read_attribute(:value)
+    YAML.load(v) unless v.nil?
+  end
+end
+
+class ConvertDnsConflictTimeoutSetting < ActiveRecord::Migration[5.2]
+  def clean_cache
+    Rails.cache.delete(Setting.cache_key("dns_conflict_timeout"))
+    Rails.cache.delete(Setting.cache_key("dns_timeout"))
+  end
+
+  def up
+    Setting.without_auditing do
+      old_setting = FakeSetting.find_by_name("dns_conflict_timeout")
+      new_setting = FakeSetting.find_by_name("dns_timeout")
+      return if old_setting.nil? || new_setting.nil?
```

Comment:
I think this can cause the issue reported in https://projects.theforeman.org/issues/27585

If the old setting was (for any reason) `nil`, this will leave the new setting also `nil` but it's expected now to be an array.

Reply:
This is indeed the bug, but not exactly for that reason: 
This is only checking if both settings exist in the database - not if their value is `nil`. Then, since we override `value` and `value=` the following code does:
```ruby
if old_setting.value != 3 # -> true, since old_setting.value == nil
  new_setting.value = [nil]
```
And dig in to `value=` we get:
```ruby
  def value=([nil])
    v = v.to_yaml unless v.nil?  # v = "---\n- \n" since [nil].nil? == false
    write_attribute :value, v # we now have "---\n- \n" in the database
  end
```

Which means that our fix for this was wrong as well - we assumed the issues was that `nil` was saved, not `[nil]`.